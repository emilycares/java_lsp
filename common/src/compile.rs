use std::process::Command;

use nom::{
    branch::alt,
    character::complete::digit0,
    multi::separated_list0,
    sequence::{pair, separated_pair},
    IResult,
};
use nom::{
    bytes::{complete::take_until, streaming::tag},
    combinator::opt,
    multi::many0_count,
};

use serde::{Deserialize, Serialize};
#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct CompileError {
    pub path: String,
    pub message: String,
    pub row: usize,
    pub col: usize,
}

pub fn compile_java_file(file_path: &str, classpath: &str) -> Option<Vec<CompileError>> {
    // Compile the Java file using `javac` with the generated classpath
    let out = Command::new("javac")
        .arg("-cp")
        .arg(classpath)
        .arg("-d")
        .arg("target/classes")
        .arg(file_path)
        .output()
        .ok()?;

    let stdout = out.stderr;
    let stdout = std::str::from_utf8(&stdout).ok()?;
    parse_compile_errors(stdout).ok().map(|e| e.1)
}

pub fn parse_compile_errors(input: &str) -> IResult<&str, Vec<CompileError>> {
    let (input, errors) = separated_list0(tag("\n"), parse_error)(input)?;
    Ok((input, errors))
}

fn parse_error(input: &str) -> IResult<&str, CompileError> {
    let (input, _) = opt(tag("\n"))(input)?;
    let (input, path) = take_until(".java:")(input)?;
    let (input, _) = tag(".java:")(input)?;
    let (input, (row, (msg, col))) =
        separated_pair(digit0, tag(": error: "), parse_message_and_col)(input)?;
    Ok((
        input,
        CompileError {
            path: format!("{}.java", path),
            message: msg.to_string(),
            row: row.parse().unwrap(),
            col,
        },
    ))
}

fn parse_message_and_col(input: &str) -> IResult<&str, (&str, usize)> {
    let (input, message) = take_until("\n")(input)?;
    let (input, _) = take_until("\n")(input)?;
    let (input, _) = tag("\n")(input)?;
    let (input, _) = take_until("\n")(input)?;
    let (input, _) = tag("\n")(input)?;
    let (input, (col, _)) = pair(many0_count(alt((tag(" "), tag("\t")))), tag("^"))(input)?;
    Ok((input, (message, col)))
}

#[cfg(test)]
mod tests {
    use crate::compile::parse_compile_errors;

    #[test]
    fn parse_compile_errors_basic() {
        let input = "
src/main/java/org/acme/GreetingResource.java:15: error: > or ',' expected
        var hash = new HashMap<String, String();
                                             ^

src/main/java/org/acme/GreetingResource.java:15: error: > or ',' expected
        var hash = new HashMap<String, String();
                                             ^
1 error
          ";
        let out = parse_compile_errors(input).unwrap();
        insta::assert_yaml_snapshot!(out.1);
    }

    #[test]
    fn parse_compile_errors_real() {
        let input = "/home/emily/Documents/java/getting-started/src/main/java/org/acme/GreetingResource.java:16: error: > or ',' expected\n\tvar hash = new HashMap<String, String();\n\t                                     ^\n1 error\n";
        let out = parse_compile_errors(input).unwrap();
        insta::assert_yaml_snapshot!(out.1);
    }
}
