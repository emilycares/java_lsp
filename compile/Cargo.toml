[package]
name = "compile"
version.workspace = true
edition.workspace = true
authors.workspace = true
homepage.workspace = true
repository.workspace = true
documentation.workspace = true
readme.workspace = true
license.workspace = true

[dependencies]
nom.workspace = true

[dev-dependencies]
pretty_assertions.workspace = true
