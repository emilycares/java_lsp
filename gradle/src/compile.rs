use std::process::Command;

use common::compile::{parse_compile_errors, CompileError};

pub fn compile_java() -> Option<Vec<CompileError>> {
    if let Some(log) = run_compile_java() {
        if let Some(value) = cut_and_parse(log) {
            return Some(value);
        }
    }

    None
}

fn cut_and_parse(log: String) -> Option<Vec<CompileError>> {
    let log = cut_log(log);
    parse_compile_errors(&log).ok().map(|e| e.1)
}

pub fn cut_log(inp: String) -> String {
    let mut out = String::new();

    for line in inp.lines() {
        if line.starts_with("FAILURE:") {
            break;
        }
        out.push_str(line);
        out.push('\n');
    }
    out
}

fn run_compile_java() -> Option<String> {
    // ./gradlew dependencies --console --plain
    let output = Command::new("./gradlew")
        .arg("compileJava")
        .arg("-q")
        .output()
        .ok()?;

    Some(String::from_utf8_lossy(&output.stderr).to_string())
}

#[cfg(test)]
mod tests {
    use common::compile::CompileError;
    use pretty_assertions::assert_eq;

    use crate::compile::cut_and_parse;

    #[test]
    fn gradle_compile() {
        let inp = include_str!("../tests/compile_basic.txt");
        let out = cut_and_parse(inp.to_owned());
        assert_eq!(
            out,
            Some(vec![
                CompileError {
                    path: "/home/emily/tmp/vanilla-gradle/app/src/main/java/org/example/Other.java"
                        .to_string(),
                    message: "illegal start of type".to_string(),
                    row: 4,
                    col: 2,
                },
                CompileError {
                    path: "/home/emily/tmp/vanilla-gradle/app/src/main/java/org/example/App.java"
                        .to_string(),
                    message: "';' expected".to_string(),
                    row: 4,
                    col: 19,
                },
            ])
        );
    }
}
