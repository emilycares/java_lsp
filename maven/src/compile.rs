use std::{
    fs::{self, read_to_string},
    io::{self, Write},
    path::Path,
    process::Command,
};

use nom::{
    branch::alt,
    character::complete::{alpha0, digit0},
    multi::separated_list0,
    sequence::{pair, separated_pair},
    IResult,
};
use nom::{
    bytes::{complete::take_until, streaming::tag},
    combinator::opt,
    multi::many0_count,
};
use serde::{Deserialize, Serialize};

pub fn generate_classpath() -> Option<String> {
    let classpath_file = "target/classpath.txt";

    if Path::new(&classpath_file).exists() {
        let classpath = read_to_string(&classpath_file).ok()?;
        return Some(format!("{}:target/classes", classpath.trim()));
    }

    let output = Command::new("mvn")
        .args(&[
            "dependency:build-classpath",
            "-Dmdep.outputFile=target/classpath.txt",
        ])
        .output()
        .ok()?;

    if !output.status.success() {
        io::stderr().write_all(&output.stderr).ok()?;
        return None;
    }

    let classpath = read_to_string(&classpath_file).ok()?;
    let full_classpath = format!("{}:target/classes", classpath.trim());

    fs::write(&classpath_file, &full_classpath).ok()?;

    Some(full_classpath)
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CompileError {
    pub path: String,
    pub message: String,
    pub row: usize,
    pub col: usize,
}

pub fn compile_java_file(file_path: &str, classpath: &str) -> Option<Vec<CompileError>> {
    // Compile the Java file using `javac` with the generated classpath
    let out = Command::new("javac")
        .arg("-cp")
        .arg(classpath)
        .arg("-d")
        .arg("target/classes")
        .arg(file_path)
        .output()
        .ok()?;

    let stdout = out.stderr;
    let stdout = std::str::from_utf8(&stdout).ok()?;
    parse_compile_errors(stdout).ok().map(|e| e.1)
}

fn parse_compile_errors(input: &str) -> IResult<&str, Vec<CompileError>> {
    let (input, errors) = separated_list0(tag("\n"), parse_error)(input)?;
    Ok((input, errors))
}

fn parse_error(input: &str) -> IResult<&str, CompileError> {
    let (input, (_, path)) = pair(opt(tag("\n")), separated_list0(tag("/"), alpha0))(input)?;
    let (input, _) = take_until(".java:")(input)?;
    let (input, _) = tag(".java:")(input)?;
    let (input, (row, (msg, col))) =
        separated_pair(digit0, tag(": error: "), parse_message_and_col)(input)?;
    Ok((
        input,
        CompileError {
            path: format!("{}.java", path.join("/")),
            message: msg.to_string(),
            row: row.parse().unwrap(),
            col,
        },
    ))
}

fn parse_message_and_col(input: &str) -> IResult<&str, (&str, usize)> {
    let (input, message) = take_until("\n")(input)?;
    let (input, _) = take_until("\n")(input)?;
    let (input, _) = tag("\n")(input)?;
    let (input, _) = take_until("\n")(input)?;
    let (input, _) = tag("\n")(input)?;
    let (input, (col, _)) = pair(many0_count(alt((tag(" "), tag("\t")))), tag("^"))(input)?;
    Ok((input, (message, col)))
}

#[cfg(test)]
mod tests {
    use crate::compile::parse_compile_errors;

    #[test]
    fn parse_compile_errors_basic() {
        let input = "
src/main/java/org/acme/GreetingResource.java:15: error: > or ',' expected
        var hash = new HashMap<String, String();
                                             ^

src/main/java/org/acme/GreetingResource.java:15: error: > or ',' expected
        var hash = new HashMap<String, String();
                                             ^
1 error
          ";
        let out = parse_compile_errors(input).unwrap();
        insta::assert_yaml_snapshot!(out.1);
    }

    #[test]
    fn parse_compile_errors_real() {
        let input = "/home/emily/Documents/java/getting-started/src/main/java/org/acme/GreetingResource.java:16: error: > or ',' expected\n\tvar hash = new HashMap<String, String();\n\t                                     ^\n1 error\n";
        let out = parse_compile_errors(input).unwrap();
        dbg!(&out.0);
        insta::assert_yaml_snapshot!(out.1);
    }
}
