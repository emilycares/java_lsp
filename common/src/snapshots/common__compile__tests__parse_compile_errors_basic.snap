---
source: common/src/compile.rs
expression: out.1
snapshot_kind: text
---
- path: src/main/java/org/acme/GreetingResource.java
  message: "> or ',' expected"
  row: 15
  col: 45
- path: src/main/java/org/acme/GreetingResource.java
  message: "> or ',' expected"
  row: 15
  col: 45
